src/NuGet.Core/NuGet.Packaging/Signing/Archive/SignedPackageArchiveIOUtility.cs

```diff
@@ -47,7 +49,42 @@ public static void SeekReaderForwardToMatchByteSignature(BinaryReader reader, by
                 stream.Position += 1;
             }
 
-            throw new Exception($"Byte signature not found in zip: {BitConverter.ToString(byteSignature)}");
+            stream.Seek(offset: originalPosition, origin: SeekOrigin.Begin);
+            throw new Exception(string.Format(CultureInfo.CurrentCulture, Strings.ErrorByteSignatureNotFound, BitConverter.ToString(byteSignature)));
+        }
+
+        /// <summary>
+        /// Takes a binary reader and moves backwards the current position of it's base stream until it finds the specified signature.
+        /// </summary>
+        /// <param name="reader">Binary reader to update current position</param>
+        /// <param name="byteSignature">byte signature to be matched</param>
+        public static void SeekReaderBackwardToMatchByteSignature(BinaryReader reader, byte[] byteSignature)
+        {
+            if (reader == null)
+            {
+                throw new ArgumentNullException(nameof(reader));
+            }
+
+            var stream = reader.BaseStream;
+            var originalPosition = stream.Position;
+
+            if (originalPosition + byteSignature.Length > stream.Length)
+            {
+                throw new Exception(Strings.ErrorByteSignatureTooBig);
+            }
+
+            while (stream.Position != 0)
```

Comment:
I'm curious how perf is for this. It seems like this should use a buffer and check within the buffer instead of going byte by byte. This can be profiled later, it shouldn't be changed in this PR